README.md
# Telegram Test Api

[![npm](https://img.shields.io/npm/v/telegram-test-api.svg)](https://npm.im/telegram-test-api)
[![license](https://img.shields.io/npm/l/telegram-test-api.svg)](https://npm.im/telegram-test-api)
[![Build](https://github.com/jehy/telegram-test-api/workflows/Build/badge.svg)](https://github.com/jehy/telegram-test-api/actions/workflows/main.yml)
[![Coverage Status](https://coveralls.io/repos/github/jehy/telegram-test-api/badge.svg?branch=master)](https://coveralls.io/github/jehy/telegram-test-api?branch=master)
[![dependencies Status](https://david-dm.org/jehy/telegram-test-api/status.svg)](https://david-dm.org/jehy/telegram-test-api)
[![devDependencies Status](https://david-dm.org/jehy/telegram-test-api/dev-status.svg)](https://david-dm.org/jehy/telegram-test-api?type=dev)
[![Known Vulnerabilities](https://snyk.io/test/github/jehy/telegram-test-api/badge.svg)](https://snyk.io/test/github/jehy/telegram-test-api)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/jehyrus)

This is telegram's web server emulator.

It is designed for testing telegram bots without using actual telegram server.

You can either include it your Node.JS test code or start api separately as a service.

Client requests to API can be made via special client, received from code, or via special api (not compatible with telegram client api).

## Using as a separate service

1. `git clone https://github.com/jehy/telegram-test-api.git && cd telegram-test-api`
2. `cp config/config.sample.json config/config.json` and edit `config/config.json` if you need
3. run `npm start` to start server
3. Take a look at `src/modules/telegramClient` to emulate a client in any 
language you want via simple HTTP API.
4. use standard telegram API with your own server URL to test bots.

## Using as built in module for nodejs

### 1. Installation

```bash
npm install telegram-test-api
```
### 2. Make some kind of bot that you wanna test
```js
class TestBot {
  constructor(bot) {
    bot.onText(/\/ping/, (msg, match)=> {
      let chatId = msg.from.id;
      let opts = {
        reply_to_message_id: msg.message_id,
        reply_markup: JSON.stringify({
          keyboard: [[{text: 'ok 1'}]],
        }),
      };
      bot.sendMessage(chatId, 'pong', opts);
    });
  }
}
```

### 3. Start server emulator

```js
  const TelegramServer = require('telegram-test-api');
  let serverConfig = {port: 9000};
  let server = new TelegramServer(serverConfig);
  await server.start();
```

#### Emulator options

You can pass options like this:
```js
const serverConfig = {
  "port": 9000,
  "host": "localhost",
  "storage": "RAM",
  "storeTimeout": 60
};

let server = new TelegramServer(serverConfig);
```
options:
* `port`, `host` - pretty self descriptive.
* `storeTimeout` - how many seconds you want to store user and bot messages which were not fetched
by bot or client.
* `storage` - where you want to store messages. Right now, only `RAM` option is implemented.

### 4. Make requests

#### Requests from bot

You can use any bot API which allows custom Telegram URL like this
(example for `node-telegram-bot-api`):
```js
const 
  TelegramBot    = require('node-telegram-bot-api');

  let botOptions = {polling: true, baseApiUrl: server.config.apiURL};
  telegramBot = new TelegramBot(token, botOptions);
```
Just set api url to your local instance url - and all done!

#### Requests from client

Client emulation is very easy. You can use built in client class:
```js
    let client = server.getClient(token);
    let message = client.makeMessage('/start');
    client.sendMessage(message);
    client.getUpdates();
```

Or you can take a look at `src/modules/telegramClient` and make client in any 
language you want via simple HTTP API.

### 5. Stop server

```js
await server.stop();
```

### Full sample

Your test code can look like this:
```js
const TelegramServer = require('telegram-test-api');
const TelegramBot = require('node-telegram-bot-api');

describe('Telegram bot test', () => {
  let serverConfig = {port: 9001};
  const token = 'sampleToken';
  let server;
  let client;
  beforeEach(() => {
    server = new TelegramServer(serverConfig);
    return server.start().then(() => {
      client = server.getClient(token);
    });
  });

  afterEach(function () {
    this.slow(2000);
    this.timeout(10000);
    return server.stop();
  });

  it('should greet Masha and Sasha', async function testFull() {
    this.slow(400);
    this.timeout(800);
    const message = client.makeMessage('/start');
    await client.sendMessage(message);
    const botOptions = {polling: true, baseApiUrl: server.config.apiURL};
    const telegramBot = new TelegramBot(token, botOptions);
    const testBot = new TestBot(telegramBot);
    const updates = await client.getUpdates();
    console.log(`Client received messages: ${JSON.stringify(updates.result)}`);
    if (updates.result.length !== 1) {
      throw new Error('updates queue should contain one message!');
    }
    let keyboard = JSON.parse(updates.result[0].message.reply_markup).keyboard;
    const message2 = client.makeMessage(keyboard[0][0].text);
    await client.sendMessage(message2);
    const updates2 = await client.getUpdates();
    console.log(`Client received messages: ${JSON.stringify(updates2.result)}`);
    if (updates2.result.length !== 1) {
      throw new Error('updates queue should contain one message!');
    }
    if (updates2.result[0].message.text !== 'Hello, Masha!') {
      throw new Error('Wrong greeting message!');
    }
    return true;
  });
});
```


CHANGELOG.MD
# Changelog

## v4

* 4.2.1 Fix starting server with `npm start`
* 4.2.0 Bump deps, add editMessageReplyMarkup method support (thanks to https://github.com/AlexxNB )
* 4.1.3 Fixed default export for non-TS usage
* 4.1.2 Added support for the rest of message fields (e.g. location)
* 4.1.1 Fixed EventEmitter leakage
* 4.1.0
  * added support for `/editMessageText`, which is necessary to leverage all the power of inline keyboards
  * added support for `/answerCallbackQuery`, which is necessary to remove spinner from inline keyboards
  * fix `client.getUpdates()` so that it returns and mark as read only updates intended for current client (chat)
* 4.0.0 Refactor using typescript. No external changes expected, but everything is possible :)

## v3
* 3.0.0 `server.ApiURL` moved to `server.config.apiURL`

## v2
* 2.6.0
  * Make webhooks work with user's commands and callback queries;
  * bump deps;
  * Remove bluebird from deps;
* 2.5.0 Added webhook support;
* 2.4.6 fix storage cleanup;
* 2.4.5 Update dependencies;
* 2.4.4 Update dependencies;
* 2.4.3 Update dependencies;
* 2.4.2 Fix [hangs issue](https://github.com/jehy/telegram-test-api/issues/20);
* 2.4.1 Added [support for https](https://github.com/jehy/telegram-test-api/issues/21);
* 2.4.0 Simple sendCallback implementation;
* 2.3.1 Add chatTitle client option;
* 2.3.0 Added Add `deleteWebhook` dummi API command, bumped deps;
* 2.2.1 fix accidental coverage issue;
* 2.2.0 added client method `makeCommand` and `sendCommand` to send commands instead of pure text;
* 2.1.0 added api method `getMe` and telegraf sample test, minor updates;
* 2.0.2 Better readme, minor deps updates;
* 2.0.1 Replace `lodash.merge` with `deep-extend`;
* 2.0.0 Replace `request` with `axios`, remove `ramda` from deps, make some funcs async;

## v1

* 1.2.3 Update dependencies;
* 1.2.2 Update dependencies;
* 1.2.1 Update dependencies;
* 1.2.0 `getUpdatesHistory` client method was added;
* 1.1.1 `deleteMessage` API method was added;
* 1.0.7 Implement `GET` method for api;
* 1.0.6 Bump dependencies;
* 1.0.5 Remove extra files from npm tarball;
* 1.0.4 Loosen dependencies;
* 1.0.3 Add request as a dependency;
* 1.0.2 Move logging to debug module, clear storage when stopping server;
* 1.0.1 Fix slow server stop;
* 1.0.0 Such much refactoring;
* 0.0.4 Public release;


telegramServer.js
'use strict';

const assert = require('assert');
const express = require('express');
const bodyParser = require('body-parser');
const EventEmitter = require('events');
const shutdown = require('http-shutdown');
const http = require('http');
const request = require('axios');

const debug = require('debug')('TelegramServer:server');
const debugStorage = require('debug')('TelegramServer:storage');
const sendResult = require('./modules/sendResult');
const TelegramClient = require('./modules/telegramClient');
const requestLogger = require('./modules/requestLogger');
const Routes = require('./routes/index');
const formatUpdate = require('./modules/formatUpdate');

function clone(data) {
  return JSON.parse(JSON.stringify(data));
}

class TelegramServer extends EventEmitter {
  constructor(config = {}) {
    super();
    this.config = clone(config);
    this.config.port = this.config.port || 9000;
    this.config.host = this.config.host || 'localhost';
    this.config.protocol = this.config.protocol || 'http';
    this.ApiURL = `${this.config.protocol}://${this.config.host}:${this.config.port}`;
    this.config.storage = this.config.storage || 'RAM';
    this.config.storeTimeout = (this.config.storeTimeout || 60) * 1000; // store for a minute by default
    debug(`Telegram API server config: ${JSON.stringify(this.config)}`);

    this.updateId = 1;
    this.messageId = 1;
    this.callbackId = 1;
    this.webhooks = {};
    this.webServer = express();
    this.webServer.use(sendResult);
    this.webServer.use(bodyParser.json());
    this.webServer.use(bodyParser.urlencoded({extended: true}));
    this.webServer.use(express.static('public'));
    this.webServer.use(requestLogger);

    if (this.config.storage === 'RAM') {
      this.storage = {
        userMessages: [],
        botMessages: [],
      };
    }
    this.started = false;
  }

  getClient(botToken, options) {
    // console.log(this);
    return new TelegramClient(this.ApiURL, botToken, options);
  }

  addBotMessage(message, botToken) {
    const d = new Date();
    const millis = d.getTime();
    const add = {
      time: millis,
      botToken,
      message,
      updateId: this.updateId,
      messageId: this.messageId,
      isRead: false,
    };
    this.storage.botMessages.push(add);
    this.messageId++;
    this.updateId++;
    this.emit('AddedBotMessage');
  }

  async waitBotMessage() {
    return new Promise((resolve) => this.on('AddedBotMessage', () => resolve()));
  }

  async waitUserMessage() {
    return new Promise((resolve) => {
      this.on('AddedUserMessage', () => resolve());
      this.on('AddedUserCommand', () => resolve());
      this.on('AddedUserCallbackQuery', () => resolve());
    });
  }

  async addUserMessage(message) {
    await this.addUserUpdate(message, {message});
    this.messageId++;
    this.emit('AddedUserMessage');
  }

  async addUserCommand(message) {
    assert.ok(message.entities, 'Command should have entities');
    await this.addUserUpdate(message, {message, entities: message.entities});
    this.messageId++;
    this.emit('AddedUserCommand');
  }

  async addUserCallback(callbackQuery) {
    await this.addUserUpdate(callbackQuery, {callbackQuery, callbackId: this.callbackId});
    this.callbackId++;
    this.emit('AddedUserCallbackQuery');
  }

  /** @private */
  async addUserUpdate(message, updateFields) {
    assert.ok(message.botToken, 'The message must be of type object and must contain `botToken` field.');
    const d = new Date();
    const millis = d.getTime();
    const add = {
      time: millis,
      botToken: message.botToken,
      updateId: this.updateId,
      messageId: this.messageId,
      isRead: false,
      ...updateFields,
    };
    this.updateId++;
    const webhook = this.webhooks[message.botToken];
    if (webhook) {
      const options = {
        url: webhook.url,
        method: 'POST',
        data: formatUpdate(add),
      };
      const resp = await request(options);
      if (resp.status > 204) {
        debug(`Webhook invocation failed: ${JSON.stringify({
          url: webhook.url,
          method: 'POST',
          requestBody: add,
          responseStatus: resp.status,
          responseBody: resp.body,
        })}`);
        throw new Error('Webhook invocation failed');
      }
    } else {
      this.storage.userMessages.push(add);
    }
  }

  messageFilter(message) {
    const d = new Date();
    const millis = d.getTime();
    return message.time > millis - this.config.storeTimeout;
  }

  cleanUp() {
    debugStorage('clearing storage');
    debugStorage(`current userMessages storage: ${this.storage.userMessages.length}`);
    this.storage.userMessages = this.storage.userMessages.filter(this.messageFilter, this);
    debugStorage(`filtered userMessages storage: ${this.storage.userMessages.length}`);
    debugStorage(`current botMessages storage: ${this.storage.botMessages.length}`);
    this.storage.botMessages = this.storage.botMessages.filter(this.messageFilter, this);
    debugStorage(`filtered botMessages storage: ${this.storage.botMessages.length}`);
  }

  cleanUpDaemon() {
    if (!this.started) {
      return;
    }
    this.cleanUpDaemonInterval = setInterval(this.cleanUp.bind(this), this.config.storeTimeout);
  }

  /**
   * Obtains all updates (messages or any other content) sent or received by specified bot.
   * Doesn't mark updates as "read".
   * Very useful for testing `deleteMessage` Telegram API method usage.
   */
  getUpdatesHistory(token) {
    return this.storage.botMessages.concat(this.storage.userMessages)
      .filter((item)=>item.botToken === token)
      .sort((a, b)=>a.time - b.time);
  }

  async start() {
    const app = this.webServer;
    const self = this;
    for (let i = 0; i < Routes.length; i++) {
      Routes[i](app, self);
    }
    // there was no route to process request
    app.use((req, res) => {
      res.sendError(new Error('Route not found'));
    });
    // Catch express bodyParser error, like http://stackoverflow.com/questions/15819337/catch-express-bodyparser-error
    app.use((error, req, res) => {
      debug(`Error: ${error}`);
      res.sendError(new Error(`Something went wrong. ${error}`));
    });
    self.server = http.createServer(app);
    self.server = shutdown(self.server);
    await new Promise((resolve, reject) => {
      self.server.listen(self.config.port, self.config.host)
        .once('listening', resolve)
        .once('error', reject);
    });
    debug(`Telegram API server is up on port ${self.config.port} in ${app.settings.env} mode`);
    self.started = true;
    self.cleanUpDaemon();
  }

  removeUserMessage(updateId) {
    this.storage.userMessages = this.storage.userMessages
      .filter((update) => (update.updateId !== updateId));
  }

  removeBotMessage(updateId) {
    this.storage.botMessages = this.storage.botMessages
      .filter((update) => update.updateId !== updateId);
  }

  setWebhook(webhook, botToken) {
    this.webhooks[botToken] = webhook;
    debug(`Webhook for bot ${botToken} set to: ${webhook.url}`);
  }

  deleteWebhook(botToken) {
    this.webhooks[botToken] = undefined;
    debug(`Webhook unset for bot ${botToken}`);
  }

  /**
   * Deletes specified message from the storage: sent by bots or by clients.
   * @returns {boolean} - `true` if the message was deleted successfully.
   */
  deleteMessage(chatId, messageId) {
    const isMessageToDelete = (update) => (
      update.message.chat.id === chatId && update.messageId === messageId
    );
    const userUpdate = this.storage.userMessages.find(isMessageToDelete);

    if (userUpdate) {
      this.removeUserMessage(userUpdate.updateId);
      return true;
    }

    const botUpdate = this.storage.botMessages.find(isMessageToDelete);

    if (botUpdate) {
      this.removeBotMessage(botUpdate.updateId);
      return true;
    }

    return false;
  }

  close() {
    this.storage = {
      userMessages: [],
      botMessages: [],
    };
  }

  async stop() {
    if (this.server === undefined) {
      debug('Cant stop server - it is not running!');
      return false;
    }
    this.started = false;
    if (this.cleanUpDaemonInterval) {
      clearInterval(this.cleanUpDaemonInterval);
    }

    const expressStop = new Promise((resolve) => {
      this.server.shutdown(() => {
        resolve();
      });
    });
    debug('Stopping server...');
    this.close();
    await expressStop;
    debug('Server shutdown ok');
    return true;
  }
}

module.exports = TelegramServer;
